lix/packages/client/src/git/pull.ts

```diff
@@ -33,47 +56,72 @@ export async function pull(
 		throw new Error("could not fetch head")
 	}
 
-	await isoGit.merge({
-		fs: pullFs,
-		cache: ctx.cache,
-		dir: ctx.dir,
-		ours: state.branchName,
-		theirs: fetchHead,
-		fastForward: cmdArgs.fastForward,
-		message: `Merge ${fetchHeadDescription}`,
-		author: cmdArgs.author || ctx.author,
-		dryRun: false,
-		noUpdateBranch: false,
-		// committer,
-		// signingKey,
-		// fastForwardOnly,
-	})
+	let materialized: string[] = []
 
+	// FIXME: there is still a race condition somewhere that can lead to materialized files being deleted
 	if (ctx.experimentalFeatures.lazyClone) {
-		console.warn(
-			"enableExperimentalFeatures.lazyClone is set for this repo but pull not fully implemented. disabling lazy files"
-		)
-
-		await emptyWorkdir(ctx, state)
+		materialized = await emptyWorkdir(ctx, state)
+		ctx.debug && console.info("experimental checkout after pull preload:", materialized)
+		state.checkedOut.clear()
+	}
 
-		// remember we are now leaving lazy mode
-		ctx.experimentalFeatures.lazyClone = false
+	/**
+	 * @typedef {Object} MergeDriverParams
+	 * @property {Array<string>} branches
+	 * @property {Array<string>} contents
+	 * @property {string} path
+	 */
+	/**
+	 * @callback MergeDriverCallback
+	 * @param {MergeDriverParams} args
+	 * @return {{cleanMerge: boolean, mergedText: string} | Promise<{cleanMerge: boolean, mergedText: string}>}
+	 */
+	const mergeDriver = ({
+		branches,
+		contents,
+		path,
+	}: {
+		branches: string[]
+		contents: string[]
+		path: string
+	}) => {
+		console.info("mergeDriver", branches, contents, path)
+		ctx.rawFs.writeFile(path + `.${branches[2]!.slice(0, 4)}.conflict`, contents[2] || "")
 
-		ctx.debug && console.info('checking out "HEAD" after pull')
+		return { cleanMerge: true, mergedText: contents[1] || "" }
+	}
 
-		await doCheckout({
-			fs: ctx.rawFs,
+	const mergeRes = await isoGit
+		.merge({
+			fs: state.nodeishFs,
 			cache: ctx.cache,
 			dir: ctx.dir,
-			ref: state.branchName,
-			noCheckout: false,
+			ours: branchName,
+			theirs: fetchHead,
+			fastForward: cmdArgs.fastForward === false ? false : true,
+			message: `Merge ${fetchHeadDescription}`,
+			author: cmdArgs.author || ctx.author,
+			dryRun: false,
+			noUpdateBranch: false,
+			abortOnConflict: true,
+			mergeDriver,
+			// committer,
+			// signingKey,
+			// fastForwardOnly,
 		})
+		.catch((error) => ({ error }))
+
+	// @ts-ignore
+	console.info("mergeRes", { data: mergeRes.data, code: mergeRes.code, error: mergeRes.error })
+
+	if (ctx.experimentalFeatures.lazyClone) {
+		await checkOutPlaceholders(ctx, state, { preload: materialized })
```

Comment:
We need to add ignorfiles to preload that have been introduced in the commits we just pulled